src/planner.c

```diff
@@ -510,21 +577,47 @@ classify_relation(const PlannerInfo *root, const RelOptInfo *rel, Hypertable **p
 				reltype = TS_REL_HYPERTABLE;
 			else
 			{
-				/* This case is hit also by non-chunk BASERELs and might slow
-				 * down planning since it requires a metadata scan every
-				 * time. But there's probably no way around it if we are to
-				 * reliably identify chunk BASERELs. We should, however, be able
-				 * to identify these in the query preprocessing and cache them
-				 * there if we need to speed this up. */
-				Chunk *chunk = ts_chunk_get_by_relid(rte->relid, false);
-
-				if (chunk != NULL)
+				/*
+				 * This case is hit also by non-chunk BASERELs. We need a costly
+				 * chunk metadata scan to distinguish between chunk and non-chunk
+				 * baserel, so we cache the result of this lookup to avoid doing
+				 * it repeatedly.
+				 *
+				 * First, check if this reloid is in cache.
+				 */
+				bool found = false;
+				ChunkCacheEntry *entry = ChunkCache_insert(ts_chunk_cache, rte->relid, &found);
+				if (found)
+				{
+					ht = entry->ht;
+					reltype = entry->type;
+					break;
+				}
+
+				/*
+				 * This reloid is not in the chunk cache, so do the full metadata
+				 * lookup.
+				 */
+				Oid hypertable_id = ts_chunk_get_hypertable_id_by_relid(rte->relid);
+				if (hypertable_id != InvalidOid)
 				{
+					/*
+					 * This is a chunk. Look up the hypertable for it.
+					 */
 					reltype = TS_REL_CHUNK;
-					ht = get_hypertable(chunk->hypertable_relid, CACHE_FLAG_NONE);
+
+					Oid hypertable_relid = ts_hypertable_id_to_relid(hypertable_id);
+					ht = get_hypertable(hypertable_relid, CACHE_FLAG_NONE);
```

Comment:
This is allocated in the `CacheMemoryContext`, but the `ts_chunk_cache` is the `CurrentMemoryContext` and there is no guarantee that they are the same.

Reply:
These are unrelated as we discussed above.